### PR TITLE
Introduce interval class for hit() methods

### PR DESCRIPTION
**Pull request by hollasch**
Lots of additional use cases for interval pending, but just tackling the `hit` methods for now.

### CODE REVIEW

File: books/RayTracingInOneWeekend.html [hunk 1]
```diff
@@ -1278,6 +1278,128 @@
...
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
+    bool sphere::hit(const ray& r, interval ray_t, hit_record& rec) const {
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+        ...
+
+        // Find the nearest root that lies in the acceptable range.
+        auto root = (-half_b - sqrtd) / a;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++ highlight
+        if (!t.contains(root)) {
```
Comment (hollasch) [inline books/RayTracingInOneWeekend.html:1376]: Will fix this in subsequent commit.

File: books/RayTracingTheNextWeek.html [hunk 2]
```diff
@@ -662,14 +661,14 @@
...
-            bool hit(const ray& r, double ray_tmin, double ray_tmax) const {
+            bool hit(const ray& r, interval ray_t) const {
                 for (int a = 0; a < 3; a++) {
                     auto t0 = fmin((minimum[a] - r.origin()[a]) / r.direction()[a],
                                    (maximum[a] - r.origin()[a]) / r.direction()[a]);
                     auto t1 = fmax((minimum[a] - r.origin()[a]) / r.direction()[a],
                                    (maximum[a] - r.origin()[a]) / r.direction()[a]);
-                    ray_tmin = fmax(t0, ray_tmin);
-                    ray_tmax = fmin(t1, ray_tmax);
+                    auto ray_tmin = fmax(t0, ray_t.min);
```
Comment (hollasch) [inline books/RayTracingTheNextWeek.html:670]: Future change using interval intersection.

**Comment by trevordblack** [comment]
CHANGELOG needs addition

**Comment by hollasch** [comment]
_Oif_.

**Comment by trevordblack** [comment]
What am I looking at?

this seems like a gargantuan omni-bus of a PR

**Comment by hollasch** [comment]
Yup. Basically, introduce the `interval` class, and then deploy its first use case: `hittable::hit` methods.

**Comment by hollasch** [comment]
I quickly identified at least eight different cases where this would help, and I'd rather develop/PR them one-by-one than submit one gigantic hairball.

**Comment by hollasch** [comment]
Ref: #777

**Review by trevordblack — APPROVED** [review]
LGTM

File: src/common/interval.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#ifndef INTERVAL_H
+#define INTERVAL_H
+//==============================================================================================
+// To the extent possible under law, the author(s) have dedicated all copyright and related and
+// neighboring rights to this software to the public domain worldwide. This software is
+// distributed without any warranty.
+//
+// You should have received a copy (see file COPYING.txt) of the CC0 Public Domain Dedication
+// along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//==============================================================================================
```
[the remark this answers is not in the document]
Comment (hollasch) [reply src/common/interval.h:10]: Thanks for the review. What about the copyright? It's the same as the others, except for the "Originally written" lines some have.

**Pull request merged**